### PR TITLE
feature: Add timeouts for handshake functions

### LINKED ISSUE #7270: Add timeouts for handshake functions:
### Describe the feature

These functions wait (indefinitely) for incoming data from the remote.
This can cause issues if the remote peer never sends data and should be mitigated with a tokio::time::timeout wrapper like

```rust
tokio::time::timeout(HANDSHAKE_TIMEOUT, self.inner.next()).await
```

we need a new error variant for this and new tests

ref:

https://github.com/paradigmxyz/reth/blob/66c9403ea1d3d03938eafc099e64e789aadc3616/crates/net/eth-wire/src/ethstream.rs#L71

https://github.com/paradigmxyz/reth/blob/66c9403ea1d3d03938eafc099e64e789aadc3616/crates/net/ecies/src/stream.rs#L54

https://github.com/paradigmxyz/reth/blob/66c9403ea1d3d03938eafc099e64e789aadc3616/crates/net/ecies/src/stream.rs#L80

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JackG-eth**
closes #7270 

These functions wait (indefinitely) for incoming data from the remote.
This can cause issues if the remote peer never sends data and should be mitigated with a tokio::time::timeout wrapper like

### CODE REVIEW

**Comment by JackG-eth** [comment]
Currently in draft as i'd like some guidance:

1. Should ecies timeout const be moved to somewhere else? if so where, or should we reference p2pstream handshake error.
2. I noticed my clippy added a lot of ';' should i remove this? if so what is good practice in this case?
3. Is mapping the error the right way to go here, or should we be matching?

Thanks!

File: crates/net/ecies/src/stream.rs [hunk 1]
```diff
@@ -51,7 +54,9 @@ where
 
         trace!("waiting for ecies ack ...");
 
-        let msg = transport.try_next().await?;
+        let msg = tokio::time::timeout(HANDSHAKE_TIMEOUT, transport.try_next())
+            .await
+            .map_err(|_| ECIESErrorImpl::StreamTimeout)??;
```
Comment (mattsse) [inline crates/net/ecies/src/stream.rs:59]: I'd like to move this to a new function
connect_with timeout, that applies a wrapper around connect, this way we only need the timeout once

File: crates/net/eth-wire/src/ethstream.rs [hunk 2]
```diff
@@ -66,19 +67,21 @@ where
             .send(alloy_rlp::encode(ProtocolMessage::from(EthMessage::Status(status))).into())
             .await?;
 
-        let their_msg_res = self.inner.next().await;
+        let their_msg_res = tokio::time::timeout(HANDSHAKE_TIMEOUT, self.inner.next())
+            .await
```
Comment (mattsse) [inline crates/net/eth-wire/src/ethstream.rs:71]: same here, we want this around the handshake function with a new function

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks, 

mind removing the trailing ;
no idea how they always appear...

File: crates/net/ecies/src/stream.rs (hunk 1 above)
Comment (JackG-eth) [reply crates/net/ecies/src/stream.rs]: Have abstracted so that handshake is wrapped by a timeout function, i hope i understood this request.

**Review by mattsse — CHANGES_REQUESTED** [review]
no breaking changes for the existing API please,

the existing API should use the default constant, connect_with_timeout should accept a timeout as parameter

**Comment by JackG-eth** [comment]
> no breaking changes for the existing API please,
> 
> the existing API should use the default constant, connect_with_timeout should accept a timeout as parameter

should be fixed.

**Review by mattsse — APPROVED** [review]
lgtm,

perhaps @Rjected was better idea for an alternative name to `connect_no_timeout`?

File: crates/net/ecies/src/stream.rs [hunk 3]
```diff
@@ -40,6 +46,27 @@ where
...
+        remote_id: PeerId,
+        timeout_limit: Duration,
+    ) -> Result<Self, ECIESError> {
+        timeout(timeout_limit, Self::connect_no_timeout(transport, secret_key, remote_id))
+            .await
+            .map_err(|_| ECIESError::from(ECIESErrorImpl::StreamTimeout))?
+    }
+
+    /// Connect to an `ECIES` server with no timeout.
+    pub async fn connect_no_timeout(
```
Comment (Rjected) [inline crates/net/ecies/src/stream.rs:66]: `without` seems slightly better to me, can't think of anything more concise though:

```suggestion
    /// Connect to an `ECIES` server with no timeout.
    pub async fn connect_without_timeout(
```

**Review by Rjected — APPROVED** [review]
lgtm, left a naming suggestion

File: crates/net/ecies/src/stream.rs (hunk 3 above)
Comment (JackG-eth) [reply crates/net/ecies/src/stream.rs]: agree, have changed for handshake as well.

**Pull request merged**